README.md
# <img src="./doc/logo.png" alt="logo" width="50"/> TRON-CLI

TRON-CLI is a command line tool that allows developers to quickly set up and manage their TRON nodes and grid api service.

```
 _________  ____  _  __    _______   ____
/_  __/ _ \/ __ \/ |/ /___/ ___/ /  /  _/
 / / / , _/ /_/ /    /___/ /__/ /___/ /  
/_/ /_/|_|\____/_/|_/    \___/____/___/
```

| Python | JDK |
|--------|-----|
| 3.6+   | Oracle 1.8 |

* Learn more about tron on [TRON Developer Hub](https://developers.tron.network/docs/full-node)

* Join the community on [TRON Discord](https://discord.gg/GsRgsTD)

* Source code on [Github](https://github.com/tronprotocol/tron-cli)

* Project on [Pypi](https://pypi.org/project/troncli/)

------

## Install

### [pip](https://pypi.org/project/troncli/)

> pip install --upgrade pip

```
pip install troncli
```

------

## Usage

### Interactive Mode

```
tron-cli i
```

Quickly set up what you want by answerinig a few questions.

![gif](./doc/imode.gif)

### Quick Command

```
tron-cli quick
```

By default, it will set up a private test-net full node for you. 

* You can also ```--nettype main``` to set up a full node which syncs to main net.

* ```--reset True``` to reset all

![gif](./doc/quick.gif)

### Advanced Set up

You can set up a full node, solidity node, event node, and grid api service (local tron-grid) with more detailed configurations via the provided subcommands. See usage by ```tron-cli -h```, and ```-h``` on each subcommands.

#### Manual 

| Command                                                                                                                                                                                                                            | Functions                          | Example1         | Example2                                                                                                                                                                                                                                                                                                                        |
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| tron-cli init --version --reset                                                                                                                                                                                                    | Init dirs and fetch code.          | tron-cli init    | tron-cli init --version 3.2.2 --reset True                                                                                                                                                                                                                                                                                      |
| tron-cli config --nettype ---nettype --fullhttpport --solhttpport --eventhttpport --fullrpcport --solrpcport --eventrpcport --enablememdb --dbsyncmode --saveintertx --savehistorytx --gridport --dbname --dbusername --dbpassword | Create and customize config files. | tron-cli config  | tron-cli config --nettype private --fullhttpport 8500 --solhttpport 8600 --eventhttpport 8400 --fullrpcport 58500 --solrpcport 58600 --eventrpcport 58400 --enablememdb True --dbsyncmode async --saveintertx False --savehistorytx False --gridport 18891 --dbname events --dbusername tron --dbpassword 12345678 --reset True |
| tron-cli run --nodetype                                                                                                                                                                                                            | Run node.                          | tron-cli run     | tron-cli run --nodetype full                                                                                                                                                                                                                                                                                                    |
| tron-cli stop --node                                                                                                                                                                                                               | Stop node.                         | tron-cli stop    | tron-cli stop --node 7777                                                                                                                                                                                                                                                                                                       |
| tron-cli status --node                                                                                                                                                                                                             | Monitor nodes status.              | tron-cli status  | tron-cli status --node 777                                                                                                                                                                                                                                                                                                      |
| tron-cli quick --reset                                                                                                                                                                                                             | Quick start.                       | tron-cli quick   | tron-cli quick --nettype main --reset True                                                                                                                                                                                                                                                                                      |
| tron-cli log --nodetype --filter                                                                                                                                                                                                   | Show filtered log.                 | tron-cli log     | tron-cli --nodetype sol --filter height                                                                                                                                                                                                                                                                                         |
| tron-cli i                                                                                                                                                                                                                         | Switch to Interactive Mode.        | tron-cli i       | tron-cli i                                                                                                                                                                                                                                                                                                                      |
| tron-cli version                                                                                                                                                                                                                   | Check installed troncli version.   | tron-cli version | tron-cli version                                                                                                                                                                                                                                                                                                                |
| tron-cli -h, --help                                                                                                                                                                                                                | Check help manual.                 | tron-cli -h      | tron-cli --help                                                                                                                                                                                                                                                                                                                 |

#### overall

```
tron-cli -h
```
```
usage: tron-cli [-h] {init,config,run,stop,status,quick,log,version,i} ...

which subcommand do you want?

optional arguments:
  -h, --help            show this help message and exit

subcommands:
  {init,config,run,stop,status,quick,log,version,i}
    init                Init dirs and fetch code.
    config              Create customize config files.
    run                 Run node.
    stop                Stop node.
    status              Monitor nodes status.
    quick               Quick start. (run a full private/main node by one
                        command)
    log                 Show filtered log.
    version             Check installed troncli version.
    i                   Interactive Mode.
```

##### subcommand: init

```
tron-cli init -h
```
```
usage: tron-cli init [-h] [--version VERSION] [--reset RESET]

optional arguments:
  -h, --help         show this help message and exit
  --version VERSION  specify java-tron version
  --reset RESET      reset all
```

##### subcommand: config

```
tron-cli config -h
```
```
usage: tron-cli config [-h] [--nettype NETTYPE] [--fullhttpport FULLHTTPPORT]
                       [--solhttpport SOLHTTPPORT]
                       [--eventhttpport EVENTHTTPPORT]
                       [--fullrpcport FULLRPCPORT] [--solrpcport SOLRPCPORT]
                       [--eventrpcport EVENTRPCPORT]
                       [--enablememdb ENABLEMEMDB] [--dbsyncmode DBSYNCMODE]
                       [--saveintertx SAVEINTERTX]
                       [--savehistorytx SAVEHISTORYTX] [--gridport GRIDPORT]
                       [--dbname DBNAME] [--dbusername DBUSERNAME]
                       [--dbpassword DBPASSWORD] [--reset RESET]

optional arguments:
  -h, --help            show this help message and exit
  --nettype NETTYPE     specify net type [main, private]
  --fullhttpport FULLHTTPPORT
                        specify full node http port
  --solhttpport SOLHTTPPORT
                        specify solidity node http port
  --eventhttpport EVENTHTTPPORT
                        specify event node http port
  --fullrpcport FULLRPCPORT
                        specify full node rpc port
  --solrpcport SOLRPCPORT
                        specify solidity node rpc port
  --eventrpcport EVENTRPCPORT
                        specify event node rpc port
  --enablememdb ENABLEMEMDB
                        enable/disable in memory db
  --dbsyncmode DBSYNCMODE
                        specify either db async or sync mode
  --saveintertx SAVEINTERTX
                        enable/disable save internal transcation
  --savehistorytx SAVEHISTORYTX
                        enable/disable save history transcation
  --gridport GRIDPORT   specify grid api port
  --dbname DBNAME       specify db name
  --dbusername DBUSERNAME
                        specify db user name
  --dbpassword DBPASSWORD
                        specify db password name
  --reset RESET         reset config to default settings
```

##### subcommand: run

```
tron-cli run -h
```
```
usage: tron-cli run [-h] [--nodetype NODETYPE]

optional arguments:
  -h, --help           show this help message and exit
  --nodetype NODETYPE  specify node type [full, sol, event, grid]
```

##### subcommand: stop

```
tron-cli stop -h
```
```
usage: tron-cli stop [-h] [--node NODE]

optional arguments:
  -h, --help   show this help message and exit
  --node NODE  stop node by given node id or all
```

##### subcommand: status

```
tron-cli status -h
```
```
usage: tron-cli status [-h] [--node NODE]

optional arguments:
  -h, --help   show this help message and exit
  --node NODE  check specific node detail by node id
```

##### subcommand: log

```
tron-cli log -h
```
```
usage: tron-cli log [-h] [--nodetype NODETYPE] [--filter FILTER]

optional arguments:
  -h, --help           show this help message and exit
  --nodetype NODETYPE  specify node type [full, sol, event, grid]
  --filter FILTER      specify filter [number/height]
```

##### subcommand: version

```
tron-cli version -h
```
```
usage: tron-cli version [-h]

optional arguments:
  -h, --help  show this help message and exit
```

##### subcommand: i

```
tron-cli i -h
```
```
usage: tron-cli i [-h]

optional arguments:
  -h, --help  show this help message and exit
```

------

## Common Use Cases

__Notice__: Take a look at ```tron-cli i``` first, interactive mode should cover most use cases.

#### I. set up private-net nodes

a. set up full node only

```
tron-cli quick
```

b. add a solidity node

```
tron-cli run --nodetype sol
```

#### II. set up main-net nodes

a. init

```
tron-cli init
```

b. config to main-net

```
tron-cli config --nettype main
```

c. run full node

```
tron-cli run
```

#### III. advanced config to start nodes

a. initilize

```
tron-cli init --version latest --reset True
```


CHANGELOG.md
# Changelog

#### Version 1.0.6

[X] Support java-tron version 3.5.0.1.

[X] Add support for log config.

[X] fix set up main-net syncing issue on event server;

#### Version 1.0.5

[X] I'm blushing to say, this is just for fixing a stupid typo.

#### Version 1.0.4

[X] fix set up main-net syncing issue.

#### Version 1.0.2

[X] fix enable_save_inter_tx and enable_save_history_tx config issue.

[X] simplify interact mode default selection.

#### Version 1.0.1

[X] Fix private net broadcast failure and syncing issue.

[X] Fix several typos.

[X] Support java-tron version 3.2.5.

#### Version 1.0.0

##### highlight:

* Add interactive mode.

* Check and throw error messages with insturction when tron-cli has not initialized or configed.

* Support python version 3.6.

##### details:

[X] Add check init and config in config and run, throw error and instruction if catched.

[X] Add interactive mode for whole walk through with key selections, which should cover most use cases.

[X] Fix private event node config error on local host port.

[X] quick - support pass in nettype

[X] support python version 3.6

#### Version 0.3.1

[X] config - set a default committee proposal list - enable CreationOfContracts, DelegateResource, SameTokenName, TvmTransferTrc10 for private testnet

[X] check/test/update troncli with java-tron version 3.2.4

[X] reset running node and config -- on init and quick reset

#### Version 0.3.0

##### highlight:

* Redesign config logic and provide more features.

* Add log subcommand to show log with filter.

##### details:

[X] Redesign config logic check and restore previous config settings. Only overwrite by given options.tron

[X] add config -- reset: reset to default config

[X] log - filter node logs

[X] version - show current version and instruction to upgrade


#### Version 0.2.4

[X] Fix get current node status async error.

#### Version 0.2.3

[X] Catch more errors.

#### Version 0.2.2

[X] Fix [issue#17](https://github.com/tronprotocol/tron-cli/issues/17)

[X] check java install and JDK version

[X] check python version

[X] check git install

[X] status - display recommended cpu / mem for running full nodes

[X] status & run - display log locations

[X] init & status - store and display java-tron node versions

[X] status - show test account/witness's private key and address for private net

#### Version 0.2.1

[X] store config, and show active config cmd in status

[X] provide connection info in status

[X] add --reset for quick

[X] add more instruction in command

[X] show ports info and cmd tips with run command

#### Version 0.2.0

##### highlight:

Support event-node and tron-grid set up/config/run/stop/monitor.

##### details:

[X] support java-tron 3.2.2 with more config

[X] add stop all feature

[X] update logo and change node list structure

[X] fix private net init account pk and address does not match

#### Version 0.1.6

[X] fix error on download progress bar when network not available

#### Version 0.1.5

##### highlight:

Pre-release of support event-node and tron-grid set up/config/run/stop/monitor.

##### details:

[X] init - change file structure for event-node and tron-grid

[X] init - add reset option and handler

[X] init - fetch event-node code

[X] init - fetch tron-grid code

[X] utils - make git_clone a util function

[X] utils - build util method to store more info and provide for status

[X] config - add options and handlers for event-node

[X] config - add options and handlers for tron-grid

[X] config - add event-node build and raise errors

[X] run - add event-node run

[X] run - fire up tron-grid

#### Version 0.1.4

[X] update to support version 3.2 fetch release, check version

[X] compatible check with version 3.2 release, and update config handler

[X] add more info message;

[X] provide a more neat yet detail help info and command

#### Version 0.1.3

[X] run - check single ps status

[X] keep track of all running nodes

[X] run - monitor overall system status

[X] set default value for all subcommand options

#### Version 0.1.2

[X] catch download errors

[X] add progress bar for download

[X] colorful logo and msg

[X] add more progress msg on ports config

[X] add more progress msg on net_type config

[X] add info msg type for instructions

[X] move changelog to file

[X] optimize progress bar

#### Version 0.1.0

[X] init - set up file folders, and get builds based on given version number

[X] config - init basic config file in json format, and convert to java properties format and export

[X] run - run a single main net full node

[X] quick start

[X] run - move 'run' to its handler, and async the call 

[X] stop - add sub cmd and its handler to stop all nodes (kill -15)

[X] run - change log and data store location

[X] config - add custom method to fire up private/shasta testnet

[X] config - add custom method to change port number

[X] config - add custom method to fire up solidity node

[X] run - add option to run solidity node

[X] pack to pip

[X] Doc - add more instruction in readme file

### Future Features

[ ] quick - add options to start common user cases

[ ] restart - restart all running nodes

[ ] clean up all std messages

[ ] config - set default/customize committee proposal list

[ ] log - show and filter nodes log

[ ] dump - fetch a dump

[ ] db - manage db, check and trim db

[ ] manual - add more instruction for config

[ ] interact - add interactive mode

[ ] quiet - add quiet mode

[ ] run - multiple (full) nodes

[blocked] config -- add option to config to sync with newly deployed shasta testnet


troncli/utils.py
#!/usr/bin/env python3
import os
import requests
import json
import sys
import psutil
import subprocess
import re
from colorama import Fore, Style, Back
from tqdm import tqdm

import urllib3
from troncli.constants import *

"""
Printing Messages
"""


def logo_simple():
    print(Fore.RED + Style.BRIGHT + '')
    print(' _________  ____  _  __    _______   ____')
    print('/_  __/ _ \/ __ \/ |/ /___/ ___/ /  /  _/')
    print(' / / / , _/ /_/ /    /___/ /__/ /___/ /  ')
    print('/_/ /_/|_|\____/_/|_/    \___/____/___/  ')
    print(Fore.RESET + Style.RESET_ALL + '')


def logo_shadow():
    print(Fore.RED + '')
    print('████████╗██████╗  ██████╗ ███╗   ██╗       ██████╗██╗     ██╗')
    print('╚══██╔══╝██╔══██╗██╔═══██╗████╗  ██║      ██╔════╝██║     ██║')
    print('   ██║   ██████╔╝██║   ██║██╔██╗ ██║█████╗██║     ██║     ██║')
    print('   ██║   ██╔══██╗██║   ██║██║╚██╗██║╚════╝██║     ██║     ██║')
    print('   ██║   ██║  ██║╚██████╔╝██║ ╚████║      ╚██████╗███████╗██║')
    print('   ╚═╝   ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═══╝       ╚═════╝╚══════╝╚═╝')
    print(Fore.RESET + '')


def progress_msg(content):
    print(Fore.CYAN + '[ TRON-CLI ]: ' + content + '...' + Fore.RESET)


def imode_msg(content):
    print(Back.BLUE + Fore.WHITE + Style.BRIGHT + '[ I-MODE ]: ' + Style.NORMAL + content + Fore.RESET + Back.RESET + Style.RESET_ALL)


def success_msg(content):
    print(Fore.GREEN + '✓ : ' + content + Fore.BLACK)


def warning_msg(content):
    print(Fore.YELLOW + '⚠ : ' + content)


def error_msg(content):
    print(Fore.RED + '✖ : ' + content)


def info_msg(content):
    print(Fore.MAGENTA + 'ⓘ: ' + content + Fore.RESET)


def info_msg_div():
    print(Fore.MAGENTA + '------------------' + Fore.RESET)


def status_msg(category, detail):
    if sys.stdout.isatty() and psutil.POSIX:
        fmt = '%-13s %s' % (Fore.BLUE + Style.BRIGHT + str(category),
                            Fore.RESET + Style.RESET_ALL + str(detail))
    else:
        fmt = '%-11s %s' % (category, detail)
    print(fmt)


def status_msg_div():
    print(Fore.BLUE + Style.BRIGHT + '------------------' + Fore.RESET + Style.RESET_ALL)


def msg(content):
    print(Fore.WHITE + '    ' + content + Fore.RESET)


def debug(content):
    print(Fore.YELLOW + Style.BRIGHT + 'DEBUG:  ' + content + Fore.RESET + Style.RESET_ALL)


def node_instruction():
    info_msg('Tips: ')
    info_msg('Check overall status:')
    msg('tron-cli status')
    info_msg('Check specific node status:')
    msg('tron-cli status --node <node id>')
    info_msg('Stop all nodes:')
    msg('tron-cli stop')
    info_msg('Stop specific node:')
    msg('tron-cli stop --node <node id>')


def node_cmds(node_id):
    info_msg('CMD Tips: ')
    info_msg('Check overall status:')
    msg('tron-cli status')
    info_msg('Check current node status:')
    msg('tron-cli status --node ' + str(node_id))
    info_msg('Stop all nodes:')
    msg('tron-cli stop')
    info_msg('Stop current node:')
    msg('tron-cli stop --node ' + str(node_id))


def recommendation():
    info_msg_div()
    info_msg('Hardware recommendation for running a full node: ')
    msg('CPU: 64 cores')
    msg('RAM: 64 GB')
    info_msg_div()


def log_location(root_path, node_type):
    if node_type == 'full':
        return root_path + NODES_DIR + FULL_NODE_DIR + '/logs/tron.log'
    elif node_type == 'sol':
        return root_path + NODES_DIR + SOLIDITY_NODE_DIR + '/logs/tron.log'
    else:
        return 'not recording logs'


"""
Node List
"""


class Node(object):
    def __init__(self):
        self.root_path = os.getcwd()
        # load or init node list file
        if os.path.isfile(self.root_path + '/' + RUNNING_NODE_LIST_FILE):
            phrase = Phrase()
            self.node_list = phrase.load_json_file(self.root_path + '/' + RUNNING_NODE_LIST_FILE)
        else:
            self.node_list = {'live': {'full': [], 'sol': [], 'event': [], 'grid': [], 'all': [], 'version': ''},
                              'db': {'dbname': '', 'dbusername': '', 'dbpassword': ''},
                              'config': {'nettype': 'private',
                                         'fullhttpport': 8500,
                                         'solhttpport': 8600,
                                         'eventhttpport': 8400,
                                         'fullrpcport': 58500,
                                         'solrpcport': 58600,
                                         'eventrpcport': 58400,
                                         'enablememdb': 'True',
                                         'dbsyncmode': 'async',
                                         'saveintertx': 'False',
                                         'savehistorytx': 'False',
                                         'gridport': 18891,
                                         'dbname': 'Null',
                                         'dbusername': 'Null',
                                         'dbpassword': 'Null'},
                             'init_ed': False,
                             'config_ed': False
                             }

    def get(self):
        return self.node_list

    def save(self):
        with open(self.root_path + '/' + RUNNING_NODE_LIST_FILE, 'w') as file:
            file.write(json.dumps(self.node_list))

    def reset_config(self):
        self.node_list['config'] = {'nettype': 'private',
                                    'fullhttpport': 8500,
                                    'solhttpport': 8600,
                                    'eventhttpport': 8400,
                                    'fullrpcport': 58500,
                                    'solrpcport': 58600,
                                    'eventrpcport': 58400,
                                    'enablememdb': 'True',
                                    'dbsyncmode': 'async',
                                    'saveintertx': 'False',
                                    'savehistorytx': 'False',
                                    'gridport': 18891,
                                    'dbname': 'Null',
                                    'dbusername': 'Null',
                                    'dbpassword': 'Null'}
        self.save()

    async def update_init_done(self, flag):
        self.node_list['init_ed'] = flag
        self.save()

    async def update_config_done(self, flag):
        self.node_list['config_ed'] = flag
        self.save()

    async def update_node_version(self, version):
        self.node_list['live']['version'] = version
        self.node_list['init_ed'] = True  # need to move this logic back to cli.py
        self.save()

    async def update_running_node(self, node_type, pid, execution):
        """
        node_type: "full" / "sol" / "event" / "grid"
        pid: int
        execution: "add" / "remove"
        """
        if execution == 'add':
            self.node_list['live'][node_type].append(pid)
            self.node_list['live']['all'].append(pid)
        elif execution == 'remove':
            if pid in self.node_list['live']['full']:
                self.node_list['live']['full'].remove(pid)
                self.node_list['live']['all'].remove(pid)
            elif pid in self.node_list['live']['sol']:
                self.node_list['live']['sol'].remove(pid)
                self.node_list['live']['all'].remove(pid)
            elif pid in self.node_list['live']['event']:
                self.node_list['live']['event'].remove(pid)
                self.node_list['live']['all'].remove(pid)
            elif pid in self.node_list['live']['grid']:
                self.node_list['live']['grid'].remove(pid)
                self.node_list['live']['all'].remove(pid)
            else:
                warning_msg('process id: ' + str(pid) + ' not in the running node list')
        else:
            error_msg('wrong execution key word: ' + str(execution))

        self.save()
        # with open(self.root_path + '/' + RUNNING_NODE_LIST_FILE, 'w') as file:
        #      file.write(json.dumps(self.node_list))

    async def update_db_settings(self, dbname, dbusername, dbpassword):
        self.node_list['db']['dbname'] = dbname
        self.node_list['db']['dbusername'] = dbusername
        self.node_list['db']['dbpassword'] = dbpassword

        self.save()
        # with open(self.root_path + '/' + RUNNING_NODE_LIST_FILE, 'w') as file:
        #      file.write(json.dumps(self.node_list))

    async def update_config(self, nettype, fullhttpport, solhttpport,
                            eventhttpport, fullrpcport, solrpcport, eventrpcport,
                            enablememdb, dbsyncmode, saveintertx, savehistorytx,
                            gridport, dbname, dbusername, dbpassword):
        self.node_list['config']['nettype'] = nettype
        self.node_list['config']['fullhttpport'] = fullhttpport
        self.node_list['config']['solhttpport'] = solhttpport
        self.node_list['config']['eventhttpport'] = eventhttpport
        self.node_list['config']['fullrpcport'] = fullrpcport
        self.node_list['config']['solrpcport'] = solrpcport
        self.node_list['config']['eventrpcport'] = eventrpcport
        self.node_list['config']['enablememdb'] = enablememdb
        self.node_list['config']['dbsyncmode'] = dbsyncmode
        self.node_list['config']['saveintertx'] = saveintertx
        self.node_list['config']['savehistorytx'] = savehistorytx
        self.node_list['config']['gridport'] = gridport
        self.node_list['config']['dbname'] = dbname
        self.node_list['config']['dbusername'] = dbusername
        self.node_list['config']['dbpassword'] = dbpassword

        self.node_list['config_ed'] = True

        self.save()


"""
Download
"""


async def download(file_name, url_string):
    with open(file_name, 'wb') as f:
        # remove ssl warnings
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

        try:
            resp = requests.get(url_string + '/' + file_name,
                                verify=False, stream=True)

        except OSError as err:
            # pbar.update(0)
            error_msg('OS Error -' + str(err))
            os.sys.exit()

        else:
            with tqdm(total=100) as pbar:
                total_length = resp.headers.get('content-length')
                if total_length is None:
                    pbar.update(100)
                    pbar.close()
                    f.write(resp.content)
                else:
                    _chunk_num = 10
                    _chunk_size = int(int(total_length) / _chunk_num) + 1
                    for data in resp.iter_content(chunk_size=_chunk_size):
                        f.write(data)
                        pbar.update(_chunk_num)
                    pbar.close()


async def git_clone(host, branch, tar_path):
    progress_msg('Git cloning ' + host + '-branch: ' + branch)
    cmd = 'git clone --single-branch -b ' + branch + ' ' + host
    cmd += ' ' + tar_path
    # _process = subprocess.Popen("exec " + cmd, stdout=subprocess.PIPE, shell=True)
    try:
        os.system(cmd)
    except OSError as err:
        error_msg('OS Error -' + str(err))
        os.sys.exit()


async def gradlew_build(task):
    cmd = './gradlew build -x test'
    try:
        os.system(cmd)
    except OSError as err:
        error_msg('OS Error -' + str(err))
        os.sys.exit()
    else:
        success_msg(task + ' gradlew build finished')


"""
Phrase
"""


class Phrase(object):
    @staticmethod
    def convert_bytes(n):
        symbols = ('K', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
        prefix = {}
        for i, s in enumerate(symbols):
            prefix[s] = 1 << (i + 1) * 10
        for s in reversed(symbols):
            if n >= prefix[s]:
                value = float(n) / prefix[s]
                return '%.1f%s' % (value, s)
        return "%sB" % n

    @staticmethod
    def load_json_file(json_file_path):
        f = open(json_file_path)
        _json_props = json.load(f)
        f.close()
        return _json_props

    def store_json2properties_to_file(self, json_props, target_file_path):
        """
        convert json to properties and store in target file
        """
        _properties = self.json2properties(json_props)
        _properties_str_formatted = self.properties2str(_properties)
        f = open(target_file_path, 'w')
        f.write(_properties_str_formatted)
        f.close()

    def store_json2javabeanconfig_to_file(self, json_props, target_file_path):
        """
        convert json to properties and store in target file
        """
        _properties = self.json2properties(json_props)
        _properties_str_formatted = self.properties2str_bean(_properties)
        f = open(target_file_path, 'w')
        f.write(_properties_str_formatted)
        f.close()

    @staticmethod
    def properties2str(properties_props):
        """
        convert properties to string, and change format
        """
        _formatted_str = str(properties_props)
        _formatted_str = re.sub("}, '", "},\n\n'", _formatted_str)
        _formatted_str = re.sub("':", ":", _formatted_str)
        _formatted_str = re.sub("' ", "", _formatted_str)
        _formatted_str = re.sub("'", "\"", _formatted_str)
        return _formatted_str

    @staticmethod
    def properties2str_bean(properties_props):
        """
        convert properties to string, and change format
        """
        _formatted_str = str(properties_props)
        _formatted_str = re.sub("}, '", "},\n\n'", _formatted_str)
        _formatted_str = re.sub("':", ":", _formatted_str)
        _formatted_str = re.sub("' ", "", _formatted_str)
        _formatted_str = re.sub("'", "\"", _formatted_str)
        _formatted_str = re.sub(":", " =", _formatted_str)
        _formatted_str = re.sub(", ", "\r", _formatted_str)
        _formatted_str = re.sub("\"", "", _formatted_str)
        return _formatted_str[1:-1]

    @staticmethod
    def json2properties(json_props):
        """
        Credit: this function is based on the phrase code in the project:
            echinopsii/net.echinopsii.ariane.community.cli.python3.
        """
        properties = {}
        if isinstance(json_props, list):
            for prop in json_props:
                if isinstance(prop['propertyValue'], list):
                    properties[prop['propertyName']] = prop['propertyValue'][1]

                elif isinstance(prop['propertyValue'], dict):
                    map_property = {}
                    for prop_key, prop_value in prop['propertyValue'].items():
                        if prop_value.__len__() > 1:
                            map_property[prop_key] = prop_value[1]
                        else:
                            print('json2properties - ' + prop_key +
                                  ' will be ignored as its definition is incomplete...')
                    properties[prop['propertyName']] = map_property

                elif prop['propertyType'] == 'array':
                    j_data = json.loads(prop['propertyValue'])
                    if j_data.__len__() > 1:
                        if j_data[0] == 'map':
                            t_data = []
                            for amap in j_data[1]:
                                t_data.append(DriverTools.json_map2properties(amap))
                            properties[prop['propertyName']] = t_data
                        elif j_data[0] == 'array':
                            t_data = []
                            for ar in j_data[1]:
                                t_data.append(DriverTools.json_array2properties(ar))
                            properties[prop['propertyName']] = t_data
                        else:
                            properties[prop['propertyName']] = j_data[1]
                    else:
                        print('json2properties - ' + prop['propertyName'] +
                              ' will be ignored as its definition is incomplete...')

                elif prop['propertyType'] == 'map':
                    j_data = json.loads(prop['propertyValue'])
                    map_property = DriverTools.json_map2properties(j_data)
                    properties[prop['propertyName']] = map_property

                else:
                    properties[prop['propertyName']] = prop['propertyValue']
        else:
            properties = json_props
        return properties

    @staticmethod
    def str2xml_to_file(xml_str, target_file_path):
        """
        use xml string to create logback xml file
        """
        f = open(target_file_path, 'w+')
        f.write(xml_str)